### PR TITLE
#919 - WFIRST module updates to cycle 7

### PR DESCRIPTION
**Pull request by rmandelb**
This PR has some very long-awaited updates to the WFIRST module, which Troxel has already been using off on branch #919.  They get the WCS and PSF up to Cycle 7 specifications with some added functionality, and a simplification of the PSF instantiation routine.  Some helper routines that weren't being used in practice got removed.

[This comment](https://github.com/GalSim-developers/GalSim/issues/919#issuecomment-347320054) has a list at the bottom of things that got punted for later.

I was a bit aggressive with API changes without deprecating stuff, but in practice I know that Troxel's code respects the new getPSF() already, and there are not tons of users for this submodule, so I thought it might be OK.

### CODE REVIEW

File: galsim/wfirst/wfirst_psfs.py [hunk 1]
```diff
@@ -161,7 +215,10 @@ def getPSF(SCAs=None, approximate_struts=False, n_waves=None, extra_aberrations=
 
             # In this case, we need to pad the edges of the pupil plane image, so we cannot just use
             # the stored file.
-            tmp_pupil_plane_im = galsim.fits.read(galsim.wfirst.pupil_plane_file)
+            if pupil_plane_type == 'long':
+                tmp_pupil_plane_im = galsim.fits.read(galsim.wfirst.pupil_plane_file_longwave)
```
Comment (rmjarvis) [inline galsim/wfirst/wfirst_psfs.py]: Not covered.  This is the high_accuracy one, so probably don't want to actually run this in the travis runs.  Maybe can use the same trick to do an assert_raises check that will hit this along the way.

File: galsim/wfirst/wfirst_psfs.py [hunk 2]
```diff
@@ -172,269 +229,114 @@ def getPSF(SCAs=None, approximate_struts=False, n_waves=None, extra_aberrations=
...
-        if logger: logger.debug('   ... SCA %d'%SCA)
-        if wavelength is None:
-            if approximate_struts:
-                PSF = galsim.ChromaticOpticalPSF(
-                    lam=zemax_wavelength,
-                    diam=galsim.wfirst.diameter, aberrations=use_aberrations,
-                    obscuration=galsim.wfirst.obscuration, nstruts=6,
-                    oversampling=oversampling, gsparams=gsparams)
+            if pupil_plane_type == 'long':
+                pupil_plane_im = galsim.wfirst.pupil_plane_file_longwave
```
Comment (rmjarvis) [inline galsim/wfirst/wfirst_psfs.py]: Also not covered.  This one is the fastest case, so we should probably actually check it somehow.  Maybe as simple as just checking that the pupil planes generated for Z and F are different.

File: galsim/wfirst/wfirst_psfs.py [hunk 3]
```diff
@@ -172,269 +229,114 @@ def getPSF(SCAs=None, approximate_struts=False, n_waves=None, extra_aberrations=
...
-        PSF_dict[SCA]=PSF
+    # Start reading in the aberrations for that SCA
+    if logger: logger.debug('Beginning to get the PSF aberrations for SCA %d.'%SCA)
+    aberrations, x_pos, y_pos = _read_aberrations(SCA)
+    # Do bilinear interpolation, unless we're exactly at the center (default).
+    use_aberrations = _interp_aberrations_bilinear(aberrations, x_pos, y_pos, SCA_pos)
 
-    return PSF_dict
+    if extra_aberrations is not None:
+        use_aberrations += extra_aberrations
```
Comment (rmjarvis) [inline galsim/wfirst/wfirst_psfs.py]: Not covered.  Should add a test that uses this somehow.

File: galsim/wfirst/wfirst_wcs.py [hunk 4]
```diff
@@ -299,6 +249,71 @@ def getWCS(world_pos, PA=None, date=None, SCAs=None, PA_is_FPA=False):
...
+    @param PA_is_FPA If True, then the position angle that was provided was the PA of the focal
+                     plane array, not the observatory. [default: False]
+    @param tol       Tolerance for errors due to distortions, as a galsim.Angle. 
+                     [default: 0.5*galsim.arcsec]
+    @returns a CelestialCoord object indicating the center of the focal plane array.
+    """
+    if not isinstance(SCA, int):
+        raise TypeError("Must pass in an int corresponding to the SCA")
+    if not isinstance(tol, galsim.Angle):
+        raise TypeError("tol must be a galsim.Angle")
```
Comment (rmjarvis) [inline galsim/wfirst/wfirst_wcs.py]: These two aren't covered.

**Comment by matroxel** [comment]
I've changed my code to work with this new api already (unless there were further changes in the last few weeks).

**Comment by rmandelb** [comment]
@matroxel - there  were further changes in the past few days, but I specifically looked at your code to make sure it wasn't doing one of the things I was disabling (specifically, getting multiple PSFs simultaneously using getPSF()) - the call to getPSF() appeared to only be requesting a single object, so that change should not affect you.

@rmjarvis - really sorry, I meant to mention (and failed) that I was lazily planning to use the coverage report after making the PR to find the changed bits that need better unit tests.  (Is there a way to do that before the PR?)  I realized it was inadequate but I wanted codecov to tell me what to do rather than trying to figure it out myself.  ;)  Sorry for wasting your time on that bit - now that I know where to look, I'll address that in the coming days, in addition to your other comments.

**Comment by rmjarvis** [comment]
> Sorry for wasting your time on that bit 

No worries.  This was a fine task to do during a telecon that I didn't really need to pay attention to.

>  (Is there a way to do that before the PR?) 

Not easily.  That's usually what I do too.  Once you get that, a neat trick is to add lines like `raise 1`, `raise 2`, etc. where the coverage was missing.  Then add tests until you hit the TypeError for the invalid raise, at which point remove it and make sure the test still works.  Once all the invalid raises are gone, push and see if you missed anything.  It works for everything except `assert_raises(TypeError)` tests.  (For those, I often put in `raise KeyError(3)` instead.)

**Comment by rmandelb** [comment]
@rmjarvis - thanks for the review!  All your comments have been addressed, and we're back to full test coverage.

**Comment by rmandelb** [comment]
Thanks for the approval!  I will plan to merge this on the weekend unless there are objections from anybody.

**Comment by rmandelb** [comment]
@rmjarvis - do you understand why GitHub is offering to "Rebase and merge" rather than just merge?  I was going to go ahead but I didn't fully understand why that was necessary so I thought I'd check (to avoid breaking everything!)

**Comment by rmjarvis** [comment]
Some time ago, I switched it to make this our preferred way to merge.  It makes for a nicer history.  (Mostly linear, rather than a complex web of joins and splits.)  So feel free to click it.  You won't break anything.  :)

If you want to see the difference before and after, do 
```
git log --graph --oneline
```
and page down a ways.  It will be obvious when you cross over to the time before "rebase and merge".

**Comment by rmandelb** [comment]
Great, thanks!

**Comment by rmandelb** [comment]
@matroxel - please start using master branch rather than #919 .

**Pull request merged**